### DISCUSSION TITLE
Getting `Uncaught Error: undefined` when using zip function.

### DISCUSSION THREAD

**Discussion by ghost** [101arrowz/fflate#26] [Q&A] [answered]
My code is:
```js
function zipAndSend(files, ws) {
  var dir = new Object();
  for (var i = 0; i < files.length; i++) {
    dir[files[i].name] = files[i];
  }

  fflate.zip(dir, {consume: true, level: 6, mem: 8}, (err, data) => {
    ws.send(data.buffer);
  });
}
```

but when this gets called, I get `Uncaught Error: undefined.` The error seems to originate from the `d` parameter being undefined in a call to `fltn.` I basically copied the example provided for the usage of `zip`. What am I doing wrong?

**Comment by ghost** [comment]
Perhaps AsyncZippable's definition has changed, but the docs for it [here](https://github.com/101arrowz/fflate/blob/master/docs/interfaces/asynczippable.md) aren't much help.

**Reply by ghost** [reply]
Ok, It seems to me that I'm not meant to pass file objects to the zip function. I don't quite understand what I'm meant to pass to it, though. Ideally, I want to stream the file, not load the whole thing into memory.

**Comment by 101arrowz** [comment]
Wow, I didn't realize that those docs failed to generate properly. I'll see if I can fix that, I don't blame you at all for now knowing what to do.

I understand that streaming files into ZIP objects is desirable, but unfortunately the current API disallows streaming into ZIP files. Even if I added support for it, I would still need to accumulate all the compressed data into memory because I don't know how long the compressed data will be.

However, I'll solve both #24 and this together with support for streaming through a compression stream, even if it isn't possible to stream out to the ZIP. At least you won't need to load the entire file into memory, just the output, and for highly compressible but massive files, that can definitely make a difference.

For now, you need to load the files with FileReader or `.arrayBuffer()` as described in the [FAQ](https://github.com/101arrowz/fflate/wiki/FAQ)

**Reply by ghost** [reply]
Awesome, I will certainly keep an eye out for when this gets implemented.

**Comment by 101arrowz** [comment]
Update on the progress for this: I've gotten streaming ZIP creation working but I want to test it further before I push to NPM. Performance is around the same as the non-streaming version but memory usage is much lower. One of the nice things about this version is that you don't necessarily need to bring in any `fflate` compression algorithm to use it, so it has a ridiculously small bundle size for uncompressed ZIP files.

I'm also planning to add support for streaming unzipping, at which point I would consider the library to be (almost) feature-complete.

**Comment by 101arrowz** [comment, marked as answer]
@chili-b, I've added support for streaming into ZIP files as of v0.5.0. Check the documentation, ask here if you have any more questions.
